File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    stream_from 'chat_channel'
    stream_for current_account
  end

  def unsubscribed
  end

  def speak(data)
    chat = Chat.create!(message: data['message'], room_id: data['room_id'], speakable_id: data['speakable_id'], speakable_type: data['speakable_type'])
    partner_account(data['partner_id'], data['partner_type'])
    ChatChannel.broadcast_to(current_account,
      chat: chat,
      chat_time: chat.updated_at.strftime('%H:%M'),
      isCurrent_user: true
    )
    ChatChannel.broadcast_to(@partner_account,
      chat: chat,
      chat_time: chat.updated_at.strftime('%H:%M'),
      image_path:chat.speakable.image_icon_path,
      isCurrent_user: false
    )
  end

  def delete(data)
    chat = Chat.find(data['id'])
    partner_account(data['partner_id'], data['partner_type'])
    ChatChannel.broadcast_to(current_account,
      delete_chat: chat.id,
      isCurrent_user: true
    )
    ChatChannel.broadcast_to(@partner_account,
      delete_chat: chat.id,
      isCurrent_user: false
    )
    # ActionCable.server.broadcast('chat_channel', { delete_chat: chat.id })
    chat.destroy!
  end
  
  private
  
  def partner_account(partner_id, partner_type)
    case partner_type
    when 'User'
      @partner_account = User.find(partner_id)
    when 'Group'
      @partner_account = Group.find(partner_id)
    end
  end
end

File: app/javascript/channels/chat_channel.js
import consumer from './consumer'

document.addEventListener('DOMContentLoaded', () => {
  if (document.URL.match( '/chats' )){
    const chatArea = document.getElementById('chat-area');
    chatArea.scrollTo(0, chatArea.scrollHeight);

    const message = document.getElementById('message');
    let sentence =''

    const appRoom = consumer.subscriptions.create('ChatChannel', {
      connected() {
        message.addEventListener('keypress', function(e) {
          if (e.key === 'Enter') {
            if (e.target.value != '') {
              appRoom.speak(message.value, message.dataset.room, message.dataset.speakerid, message.dataset.speakertype, message.dataset.partnerid, message.dataset.partnertype);
              e.target.value = '';
            }else{
              alert('空のメッセージは送信できません');
            }
            e.preventDefault();
          }
        })
      },

      disconnected() {},

      received(data) {
        if(data['chat']){
          if (data['isCurrent_user']==true){
            sentence = `<div id='my-chat-${data['chat'].id}' class='flex items-end justify-end mt-2'>
                          <div class='flex flex-col'>
                            <a class='chat_destroy' data-remote='true' rel='nofollow' data-method='delete' href='/rooms/${data['chat'].room_id}/chats/${data['chat'].id}'>削除</a>
                            <span class='text-sm mr-1'>${data['chat_time']}</span>
                          </div>
                            <div class='chat myself'>
                              <span>${data['chat'].message}</span>
                            </div>
                        </div>`;
            
          }else{
            sentence = `<div class='flex items-end justify-start mt-4'>
                          <img class='h-6 w-6 mr-2' src='/${data['image_path']}'>
                            <div class='chat partner'>
                              <span>${data['chat'].message}</span>
                            </div>
                            <span class='text-sm ml-1'>${data['chat_time']}</span>
                          </div>`;
          }
          chatArea.insertAdjacentHTML('beforeend', sentence); 
          chatArea.scrollTo(0, chatArea.scrollHeight);
        }else if(data['delete_id']){
          return this.perform('delete', { id: data['delete_id']});
        }else if(data['delete_chat']){
          const myChat = document.getElementById(`my-chat-${data['delete_chat']}`);
          myChat.remove();
        }
      },

      speak: function(message, room, speaker_id, speaker_type, partner_id, partner_type) { 
        return this.perform('speak', { message: message, room_id: room, speakable_id: speaker_id, speakable_type: speaker_type, partner_id: partner_id, partner_type: partner_type});
      }
    });

    const chatSubmit = document.getElementById('chat-submit');
    chatSubmit.addEventListener('click', function(e) {
      if (message.value != '') {
        appRoom.speak(message.value, message.dataset.room, message.dataset.speakerid, message.dataset.speakertype, message.dataset.partnerid, message.dataset.partnertype);
        message.value = '';
      }else{
        alert('空のメッセージは送信できません');
      }
      e.preventDefault();
    })
  }
});